### PR TITLE
Add music multi-disc and boxset facility

### PR DESCRIPTION
**Pull request by the-black-eagle**
## Description
This PR adds the ability for Kodi to navigate albums by the discs inside them.  Opening an album no longer shows all the songs on that album but instead displays the 'discs' in the album.  Opening a disc then displays the songs.  If an album only has one disc, Kodi will just display the songs, as before.  The multi-disc navigation can be disabled by the user if required and Kodi will revert to current behaviour.
Further, this PR also adds automatic detection of 'box sets' and the creation of a node to display them.  As one of the criteria for a box set is that each disc has it's own title, the PR also adds the ability to filter on disc titles and the number of discs in a box set or album and also sort by the number of discs.

## Motivation and Context
With many artists releasing anniversary sets that seem to increase in size each release (10 years, 20 years etc etc) navigating them becomes increasingly difficult when the song list contains all the songs on the entire collection.  It is difficult to navigate to a particular track or indeed to easily play a single disc when there are 300+ songs in the list.  This PR therefore gives the ability to 'split' those long lists into the individual discs that correspond to the physical media you could hold in your hand.  This makes it much easier to find a specific track, eg track 4, disc 3, or to play an entire disc with one or two clicks in much the same way as we used to put a cd into a player.

As many modern multi-disc releases are 'box-sets', i.e. repackaged re-releases usually with previously unreleased material now included, there was a discussion in the forum https://forum.kodi.tv/showthread.php?tid=348047 regarding how Kodi could/should handle these.  Based on the musicbrainz tags, it was decided that box sets all have individual disc titles and contain at least 3 discs.  Kodi uses this criteria to automatically determine if an album is a box set and flags it as such in the music db if it is.  Users can also add the tag 'boxset' to the 'releasetype' tag for any album they feel should be a box set and Kodi will add it regardless of disc titles or number of discs.

When Kodi displays the discs, the disc titles read from the tags will be used if present, otherwise Kodi will generate 'disc 1, disc 2' etc in the user's language.  Individual discs inside albums can have individual thumbs and discart and the thumbs can be different from the album thumb.  Art is rotated as necessary into the current infolabels.  This PR also adds info labels to enable skins to display the disc title, the number of discs in an album/set and a boolean which is true if the current item is a boxset. 

Although this does not break any existing functionality, it does require that the user re-scan their music so that the disc titles are read into the database and the total number of discs for each album is calculated.  Therefore, the music database version has been bumped and users will be prompted to re-scan their music.

Still to do - The extra information now held in the music db needs adding to the JSON schema.  This will be done in a separate PR.

## How Has This Been Tested?

Tested multiple times on Ubuntu 18.04 with both small music libraries of 30-40 albums and also with 1000+ albums.  I also tested with both local sqlite databases and networked MySQL databases.
## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by DaveTBlake** [comment]
Thanks for this @the-black-eagle, a nice new music lib feature for v19.

Just to say to others that I have been watching over this during development, and I'm really happy with what has been done. For a first contribution to Kodi development this is great stuff. I'll kick off some test builds for the interested/enthusiastic  music users to try tomorrow.

File: xbmc/music/infoscanner/MusicInfoScanner.cpp [hunk 1]
```diff
@@ -919,6 +925,65 @@ int CMusicInfoScanner::RetrieveMusicInfo(const std::string& strDirectory, CFileI
   return numAdded;
 }
 
+void CMusicInfoScanner::CheckBoxSets(VECALBUMS& albums)
+{
+  /* Strategy: Having scanned the tags of the albums to add to the library,
+     check for disc subtitles. If an album only has one subtitled disc, we don't
+     consider it a boxset as its most likely a 'live'version of the album or a 'bonus disc'.
```
Comment (scott967) [inline xbmc/music/infoscanner/MusicInfoScanner.cpp]: Sorry for being a pedant
```suggestion
     consider it a boxset as it's most likely a 'live' version of the album or a 'bonus disc'.
```

**Comment by ksooo** [comment]
Awesome. 👍

File: xbmc/music/infoscanner/MusicInfoScanner.cpp (hunk 1 above)
Comment (the-black-eagle) [reply xbmc/music/infoscanner/MusicInfoScanner.cpp]: Fixed.

File: xbmc/guilib/guiinfo/MusicGUIInfo.cpp [hunk 2]
```diff
@@ -318,6 +325,7 @@ bool CMusicGUIInfo::GetLabel(std::string& value, const CFileItem *item, int cont
 
         value = CURL(value).GetWithoutUserDetails();
         return true;
+
```
Comment (notspiff) [inline xbmc/guilib/guiinfo/MusicGUIInfo.cpp]: meh

File: xbmc/music/MusicDatabase.cpp [hunk 3]
```diff
@@ -210,6 +212,7 @@ void CMusicDatabase::CreateTables()
   CLog::Log(LOGINFO, "create versiontagscan table");
   m_pDS->exec("CREATE TABLE versiontagscan (idVersion INTEGER, iNeedsScan INTEGER, lastscanned VARCHAR(20))");
   m_pDS->exec(PrepareSQL("INSERT INTO versiontagscan (idVersion, iNeedsScan) values(%i, 0)", GetSchemaVersion()));
+
```
Comment (notspiff) [inline xbmc/music/MusicDatabase.cpp]: meh

File: xbmc/music/MusicDatabase.cpp [hunk 4]
```diff
@@ -634,10 +650,9 @@ int CMusicDatabase::AddSong(const int idAlbum,
       idSong = m_pDS->fv("idSong").get_asInt();
       m_pDS->close();
       UpdateSong( idSong, strTitle, strMusicBrainzTrackID, strPathAndFileName, strComment, strMood, strThumb,
-                  artistDisp, artistSort, genres, iTrack, iDuration, iYear, iTimesPlayed, iStartOffset, iEndOffset,
+                  artistDisp, artistSort, genres, iTrack, iDuration, iYear, strDiscSubtitle, iTimesPlayed, iStartOffset, iEndOffset,
                   dtLastPlayed, rating, userrating, votes, replayGain);
     }
-
```
Comment (notspiff) [inline xbmc/music/MusicDatabase.cpp]: meh

File: xbmc/music/MusicDatabase.cpp [hunk 5]
```diff
@@ -4591,6 +4617,203 @@ bool CMusicDatabase::GetAlbumsByWhere(const std::string &baseDir, const Filter &
...
+      return false;
+
+    extFilter.AppendGroup("albumview.idAlbum, iDisc");
+
+    // If there are extra songview WHERE conditions adjust to song or albumview 
+    // fields, and join Path table for strPath    
+    // ! @todo: convert songview fields into to song or albumview fields
+    // But not sure we ever get songview fields in filter - REMOVE??
+    if (extFilter.where.find("songview.strPath") != std::string::npos)
+    {
```
Comment (notspiff) [inline xbmc/music/MusicDatabase.cpp]: drop braces.

File: xbmc/music/MusicDatabase.cpp [hunk 6]
```diff
@@ -4591,6 +4617,203 @@ bool CMusicDatabase::GetAlbumsByWhere(const std::string &baseDir, const Filter &
...
+      sorting.sortBy = SortByNone;       
+    if (!SortUtils::SortFromDataset(sorting, MediaTypeAlbum, m_pDS, results))
+      return false;
+
+    // Get data from returned rows, not possibly multiple albums although usually only one
+    items.Reserve(total);
+    int albumOffset = 2;
+    CAlbum album;
+    bool useTitle = false;
+    std::string oldDiscTitle = "";
```
Comment (notspiff) [inline xbmc/music/MusicDatabase.cpp]: strings are empty by default.

File: xbmc/music/MusicDatabase.cpp [hunk 7]
```diff
@@ -4591,6 +4617,203 @@ bool CMusicDatabase::GetAlbumsByWhere(const std::string &baseDir, const Filter &
...
+    const dbiplus::query_data &data = m_pDS->get_result_set().records;
+    for (const auto &i : results)
```
Comment (notspiff) [inline xbmc/music/MusicDatabase.cpp]: const auto& i

File: xbmc/music/MusicDatabase.cpp [hunk 8]
```diff
@@ -4591,6 +4617,203 @@ bool CMusicDatabase::GetAlbumsByWhere(const std::string &baseDir, const Filter &
...
+          // group discs together by title not number.
+          useTitle = true;
+          continue;
+        }
+        else
+          oldDiscTitle = strDiscSubtitle;
+
+        CMusicDbUrl itemUrl = musicUrl;
+        std::string path = StringUtils::Format("%i/", discnum);
+        //path = StringUtils::Format("musicdb://albums/%i/%i", album.idAlbum, discnum);
```
Comment (notspiff) [inline xbmc/music/MusicDatabase.cpp]: no new commented code.

File: xbmc/music/MusicDatabase.cpp [hunk 9]
```diff
@@ -4591,6 +4617,203 @@ bool CMusicDatabase::GetAlbumsByWhere(const std::string &baseDir, const Filter &
...
+        itemUrl.AppendPath(path);
+
+        // When disc titles are provided group discs together by title not number.
+        // For monster sets like https://musicbrainz.org/release/cc967f36-7e4e-4a5b-ae0d-f1a1ab2c9c5a
+        if (useTitle)
+          itemUrl.AddOption("disctitle", strDiscSubtitle.c_str());
```
Comment (notspiff) [inline xbmc/music/MusicDatabase.cpp]: unnecessary .c_str()

File: xbmc/music/MusicDatabase.cpp [hunk 10]
```diff
@@ -9016,6 +9259,38 @@ bool CMusicDatabase::GetGenresJSON(CFileItemList& items, bool bSources)
...
+  std::string albumtitle;
+  if (idAlbum > 0)
+    albumtitle = GetAlbumById(idAlbum);
+  if (idDisc > 0)
+  {
+    disctitle = GetSingleValue("song", "strDiscSubtitle",
+                               PrepareSQL("idAlbum = %i AND iTrack >> 16 = %i", idAlbum, idDisc));
+    if (disctitle.empty())
+      disctitle = StringUtils::Format("%s %i", g_localizeStrings.Get(427), idDisc); // "Disc 1" etc.
+    if (!albumtitle.empty())
```
Comment (notspiff) [inline xbmc/music/MusicDatabase.cpp]: reverse if to avoid negation

File: xbmc/music/MusicDatabase.cpp [hunk 11]
```diff
@@ -8716,6 +8952,13 @@ bool CMusicDatabase::IsSongAlbumArtist(int idSong, int idArtist)
   return GetSingleValue(strSQL).empty();
 }
 
+bool CMusicDatabase::IsAlbumBoxset(int idAlbum)
+{
+  std::string strSQL = PrepareSQL("SELECT bBoxedSet FROM album WHERE idAlbum = %i", idAlbum);
+  int isBoxSet = strtol(GetSingleValue(strSQL).c_str(), NULL, 10);
```
Comment (notspiff) [inline xbmc/music/MusicDatabase.cpp]: nullptr

File: xbmc/music/MusicDatabase.cpp [hunk 12]
```diff
@@ -9016,6 +9259,38 @@ bool CMusicDatabase::GetGenresJSON(CFileItemList& items, bool bSources)
...
+      albumtitle = albumtitle + " - " + disctitle;
+    else
+      albumtitle = disctitle;
+  }
+  return albumtitle;
+}
+
+int CMusicDatabase::GetBoxsetsCount()
+{
+  return strtol(GetSingleValue("album", "count(idAlbum)", "bBoxedSet = 1").c_str(), NULL, 10);
```
Comment (notspiff) [inline xbmc/music/MusicDatabase.cpp]: nullptr

File: xbmc/music/MusicDatabase.cpp [hunk 13]
```diff
@@ -9016,6 +9259,38 @@ bool CMusicDatabase::GetGenresJSON(CFileItemList& items, bool bSources)
...
+}
+
+int CMusicDatabase::GetAlbumDiscsCount(int idAlbum)
+{
+  std::string strSQL = PrepareSQL("SELECT iDiscTotal FROM album WHERE album.idAlbum = %i", idAlbum);
+  return strtol(GetSingleValue(strSQL).c_str(), NULL, 10);
```
Comment (notspiff) [inline xbmc/music/MusicDatabase.cpp]: nullptr

File: xbmc/music/MusicThumbLoader.cpp [hunk 14]
```diff
@@ -316,9 +321,51 @@ bool CMusicThumbLoader::FillLibraryArt(CFileItem &item)
       if (artitem.prefix == "albumartist" && artitem.artType == "fanart" && fanartfallback.empty())
         fanartfallback = artname;
     }
+
     if (!fanartfallback.empty())
       item.SetArtFallback("fanart", fanartfallback);
 
+    // Process specifc disc art when we have some
```
Comment (notspiff) [inline xbmc/music/MusicThumbLoader.cpp]: specific

File: xbmc/music/Song.cpp [hunk 15]
```diff
@@ -349,3 +351,10 @@ bool CSong::ArtMatches(const CSong &right) const
   return (right.strThumb == strThumb &&
           embeddedArt.Matches(right.embeddedArt));
 }
+
+const std::string CSong::GetDiscSubtitle() const
+{
+  if (!strDiscSubtitle.empty())
```
Comment (notspiff) [inline xbmc/music/Song.cpp]: unnecessary if'ery, just return the string.

File: xbmc/music/infoscanner/MusicInfoScanner.cpp [hunk 16]
```diff
@@ -919,6 +925,65 @@ int CMusicInfoScanner::RetrieveMusicInfo(const std::string& strDirectory, CFileI
...
+  int old_discno = 0;
+  for (auto& album : albums)
+  {
+    album.bBoxedSet = false;
+    if (album.strType.find("boxset") != std::string::npos)
+      album.bBoxedSet = true;
+    if (album.bCompilation && !album.bBoxedSet)
+      continue; // skip compilations that haven't had the boxset tag added to them
+
+    count = 0;
```
Comment (notspiff) [inline xbmc/music/infoscanner/MusicInfoScanner.cpp]: scope of var can be lessened.

File: xbmc/playlists/SmartPlayList.cpp [hunk 17]
```diff
@@ -739,6 +747,8 @@ std::string CSmartPlaylistRule::GetBooleanQuery(const std::string &negate, const
   {
     if (m_field == FieldCompilation)
       return negate + GetField(m_field, strType);
+    if (m_field == FieldIsBoxset)
+      return  negate + "albumview.bBoxedSet = 1";
```
Comment (notspiff) [inline xbmc/playlists/SmartPlayList.cpp]: one space after return.

**Review by notspiff — APPROVED** [review]
Looks good to me. All I got is nitpicks (which may or may not be addressed, hence approval).

**Comment by DaveTBlake** [comment]
Test builds of this PR are now on the mirrors e.g. for Win64 http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20191115-a35dd75e-PR16915-merge-x64.exe with OSX and Android variants in appropriate locations

File: xbmc/playlists/SmartPlayList.cpp [hunk 18]
```diff
@@ -113,7 +113,10 @@ static const translateField fields[] = {
   { "artisttype",        FieldArtistType,              CDatabaseQueryRule::TEXT_FIELD,     NULL,                                 false, 564 },
   { "gender",            FieldGender,                  CDatabaseQueryRule::TEXT_FIELD,     NULL,                                 false, 39025 },
   { "disambiguation",    FieldDisambiguation,          CDatabaseQueryRule::TEXT_FIELD,     NULL,                                 false, 39026 },
-  { "source",            FieldSource,                  CDatabaseQueryRule::TEXT_FIELD,     NULL,                                 true,  39030 }
+  { "source",            FieldSource,                  CDatabaseQueryRule::TEXT_FIELD,     NULL,                                 true,  39030 },
+  { "disctitle",         FieldDiscTitle,               CDatabaseQueryRule::TEXT_FIELD,     NULL,                                 false,  38076 },
+  { "isboxset",          FieldIsBoxset,                CDatabaseQueryRule::BOOLEAN_FIELD,  NULL,                                 false, 38074 },
+  { "totaldiscs",        FieldTotalDiscs,              CDatabaseQueryRule::NUMERIC_FIELD,  StringValidation::IsPositiveInteger,  false, 38077 }
```
Comment (Rechi) [inline xbmc/playlists/SmartPlayList.cpp]: add a comma (`,`) at the end, so further fields can be added without changing this line

File: xbmc/playlists/SmartPlayList.cpp [hunk 19]
```diff
@@ -815,6 +825,8 @@ std::string CSmartPlaylistRule::FormatWhereClause(const std::string &negate, con
       query = GetField(m_field, strType) + " is NULL or " + GetField(m_field, strType) + parameter;
     else if (m_field == FieldSource)
       query = negate + " EXISTS (SELECT 1 FROM album_source, source WHERE album_source.idAlbum = " + GetField(FieldId, strType) + " AND album_source.idSource = source.idSource AND source.strName" + parameter + ")";
+    else if (m_field == FieldDiscTitle)
+      query = negate + " EXISTS (SELECT 1 FROM song WHERE song.idAlbum = " + GetField(FieldId, strType) + " AND song.strDiscSubtitle" + parameter + ")";
```
Comment (Rechi) [inline xbmc/playlists/SmartPlayList.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/utils/DatabaseUtils.cpp [hunk 20]
```diff
@@ -69,6 +69,7 @@ std::string DatabaseUtils::GetField(Field field, const MediaType &mediaType, Dat
     else if (field == FieldDateAdded) return "albumview.dateAdded";
     else if (field == FieldPlaycount) return "albumview.iTimesPlayed";
     else if (field == FieldLastPlayed) return "albumview.lastPlayed";
+    else if (field == FieldTotalDiscs) return "albumview.iDiscTotal";
```
Comment (Rechi) [inline xbmc/utils/DatabaseUtils.cpp]: > Do not put multiple statements on a single line. Always use a new line for a new statement.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#41-multiple-statements

File: xbmc/utils/DatabaseUtils.cpp [hunk 21]
```diff
@@ -92,6 +93,7 @@ std::string DatabaseUtils::GetField(Field field, const MediaType &mediaType, Dat
     else if (field == FieldArtist || field == FieldAlbumArtist) return "songview.strArtists";
     else if (field == FieldGenre) return "songview.strGenre";
     else if (field == FieldDateAdded) return "songview.dateAdded";
+    else if (field == FieldDiscTitle) return "songview.strDiscSubtitle";
```
Comment (Rechi) [inline xbmc/utils/DatabaseUtils.cpp]: > Do not put multiple statements on a single line. Always use a new line for a new statement.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#41-multiple-statements

File: xbmc/utils/DatabaseUtils.cpp [hunk 22]
```diff
@@ -492,6 +494,7 @@ int DatabaseUtils::GetField(Field field, const MediaType &mediaType, bool asInde
     else if (field == FieldPlaycount) return CMusicDatabase::album_iTimesPlayed;
     else if (field == FieldLastPlayed) return CMusicDatabase::album_dtLastPlayed;
     else if (field == FieldDateAdded) return CMusicDatabase::album_dtDateAdded;
+    else if (field == FieldTotalDiscs) return CMusicDatabase::album_iTotalDiscs;
```
Comment (Rechi) [inline xbmc/utils/DatabaseUtils.cpp]: > Do not put multiple statements on a single line. Always use a new line for a new statement.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#41-multiple-statements

File: xbmc/utils/LabelFormatter.cpp [hunk 23]
```diff
@@ -316,6 +317,10 @@ std::string CLabelFormatter::GetMaskContent(const CMaskString &mask, const CFile
     if (music && music->GetDateAdded().IsValid())
       value = music->GetDateAdded().GetAsLocalizedDate();
     break;
+  case 'b': // Total number of discs
+    if(music)
```
Comment (Rechi) [inline xbmc/utils/LabelFormatter.cpp]: > Control statement keywords have to be separated from opening parentheses by one space.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#34-whitespace

File: xbmc/utils/SortUtils.cpp [hunk 24]
```diff
@@ -179,6 +179,11 @@ std::string ByTrackNumber(SortAttribute attributes, const SortItem &values)
   return StringUtils::Format("%i", (int)values.at(FieldTrackNumber).asInteger());
 }
 
+std::string ByTotalDiscs(SortAttribute attributes, const SortItem& values)
+{
+  return StringUtils::Format("%d %s", static_cast<int>(values.at(FieldTotalDiscs).asInteger()),
+                             ByLabel(attributes,values));
```
Comment (Rechi) [inline xbmc/utils/SortUtils.cpp]: > Commas have to be followed by one space.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#34-whitespace

File: xbmc/utils/SortUtils.cpp [hunk 25]
```diff
@@ -612,6 +617,7 @@ std::map<SortBy, SortUtils::SortPreparator> fillPreparators()
   preparators[SortByInstallDate]              = ByInstallDate;
   preparators[SortByLastUpdated]              = ByLastUpdated;
   preparators[SortByLastUsed]                 = ByLastUsed;
+  preparators[SortByTotalDiscs]               = ByTotalDiscs;
```
Comment (Rechi) [inline xbmc/utils/SortUtils.cpp]: > Do not use whitespace to vertically align around operators or values. This causes problems on code review if one needs to realign all values to their new position, producing unnecessarily large diffs.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#35-vertical-alignment

**Review by Rechi — COMMENTED** [review]
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

**Comment by fritsch** [comment]
Hint: 

```
git rebase -x 'git clang-format HEAD~' xbmc-upstream/master
```

where xbmc-upstream is tracking kodi. This will check each patch and break when clang-format changes something. Now git -a --amend followed by git rebase --continue will rebase and reformat your entire commits.

Thanks much in advance.

**Comment by the-black-eagle** [comment]
@fritsch , thanks!  It appears to have reformatted much more than just my stuff though :(
Guess I can probably rewind it.... (Hate this github stuff, it never goes right for me !!).

**Comment by DaveTBlake** [comment]
Well it does show what our current Clang format/document guidelines will do to the previously clearly layed out structures and filter definitions :(

File: xbmc/dialogs/GUIDialogMediaFilter.cpp [hunk 26]
```diff
@@ -44,88 +44,164 @@
...
-  { "songs",        FieldYear,          562,    SettingType::Integer, "range",  "integer",  CDatabaseQueryRule::OPERATOR_BETWEEN },
-  { "songs",        FieldGenre,         515,    SettingType::List,    "list",   "string",   CDatabaseQueryRule::OPERATOR_EQUALS },
-  { "songs",        FieldPlaycount,     567,    SettingType::Integer, "range",  "integer",  CDatabaseQueryRule::OPERATOR_BETWEEN },
-  { "songs",        FieldSource,      39030,    SettingType::List,    "list",   "string",   CDatabaseQueryRule::OPERATOR_EQUALS }
-};
+    {"movies", FieldTitle, 556, SettingType::String, "edit", "string",
+     CDatabaseQueryRule::OPERATOR_CONTAINS},
+    {"movies", FieldRating, 563, SettingType::Number, "range", "number",
+     CDatabaseQueryRule::OPERATOR_BETWEEN},
+    {"movies", FieldUserRating, 38018, SettingType::Integer, "range", "integer",
```
Comment (DaveTBlake) [inline xbmc/dialogs/GUIDialogMediaFilter.cpp]: I'm sure that we do not want to let Clang reformat the code like this!
@the-black-eagle this is a useful "error", but can you undo it while the team discusses how to stop Clang/change the code guidlines

Comment (the-black-eagle) [reply xbmc/dialogs/GUIDialogMediaFilter.cpp]: Yeah, I was annoyed/gutted/words I can't post ! when I saw what it did !  For ref, the command needs to be `git rebase -x 'git clang-format-6.0 HEAD~' upstream/master`.  

Anyway, I'll see what I can do to reverse it. In the meantime, please accept my apologies.  I can do it by hand if/when I've rolled it back but it'll take a few hours.  My IDE can apply the clang-formatting no problem, it's just finding the right bits of code again to apply it to.

Comment (fritsch) [reply xbmc/dialogs/GUIDialogMediaFilter.cpp]: The command works for me as is, but on my system clang-format-9 is used. Out of curiousity, before your reset everything, could you post the patchset it produced.

If it formated stuff not in your changes, something went wrong with the format.

Comment (fritsch) [reply xbmc/dialogs/GUIDialogMediaFilter.cpp]: Nevermind - I have seen it. Yes from current pov of code formating this is correct even if it looks odd. Sorry for the trouble - you hit an area where old coding style glitches very heavily with the new design rules.

Comment (Rechi) [reply xbmc/dialogs/GUIDialogMediaFilter.cpp]: If you are on a Debian based system, you have to install `clang-format` package, which depends on the default clang-format version, e.g. `clang-format-6.0` or `clang-format-9` and creates a symlink to it. That's the same as for clang, clang-tidy, gcc or g++ and many other packages.

File: xbmc/music/MusicDatabase.cpp [hunk 27]
```diff
@@ -489,6 +497,10 @@ bool CMusicDatabase::AddAlbum(CAlbum& album, int idSource)
   for (const auto &albumArt : album.art)
     SetArtForItem(album.idAlbum, MediaTypeAlbum, albumArt.first, albumArt.second);
 
+  // Set album disc total
+  m_pDS->exec(PrepareSQL("UPDATE album SET iDisctotal = (SELECT COUNT(DISTINCT iTrack >> 16) FROM song "
```
Comment (Rechi) [inline xbmc/music/MusicDatabase.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/music/MusicDatabase.h [hunk 28]
```diff
@@ -417,6 +425,12 @@ class CMusicDatabase : public CDatabase
   /////////////////////////////////////////////////
   int  GetCompilationAlbumsCount();
 
+  ////////////////////////////////////////////////
+  // Boxsets
+  ////////////////////////////////////////////////
+  bool IsAlbumBoxset(int idAlbum);
+  int  GetBoxsetsCount();
```
Comment (Rechi) [inline xbmc/music/MusicDatabase.h]: > Do not use whitespace to vertically align around operators or values. This causes problems on code review if one needs to realign all values to their new position, producing unnecessarily large diffs.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#35-vertical-alignment

File: xbmc/music/tags/TagLoaderTagLib.cpp [hunk 29]
```diff
@@ -644,6 +650,8 @@ bool CTagLoaderTagLib::ParseTag(Ogg::XiphComment *xiph, EmbeddedArt *art, CMusic
       tag.SetComment(it->second.front().to8Bit(true));
     else if (it->first == "CUESHEET")
       tag.SetCueSheet(it->second.front().to8Bit(true));
+     else if (it->first == "DISCSUBTITLE")
```
Comment (Rechi) [inline xbmc/music/tags/TagLoaderTagLib.cpp]: > Use spaces as tab policy with an indentation size of 2.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#32-indentation

File: xbmc/music/tags/TagLoaderTagLib.cpp [hunk 30]
```diff
@@ -1062,6 +1072,14 @@ void CTagLoaderTagLib::AddArtistRole(CMusicInfoTag &tag, const std::string& strR
     tag.AddArtistRole(strRole, values);
 }
 
+void CTagLoaderTagLib::SetDiscSubtitle(CMusicInfoTag& tag, const::std::vector<std::string>& values)
```
Comment (Rechi) [inline xbmc/music/tags/TagLoaderTagLib.cpp]: `const std::vector<std::string>&`

File: xbmc/playlists/SmartPlayList.cpp (hunk 17 above)
Comment (Rechi) [reply xbmc/playlists/SmartPlayList.cpp]: > Do not use whitespace to vertically align around operators or values. This causes problems on code review if one needs to realign all values to their new position, producing unnecessarily large diffs.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#35-vertical-alignment

**Review by Rechi — COMMENTED** [review]
Please adjust the commit description. It contains the commit messages which you have squashed into a single one.

File: xbmc/guilib/guiinfo/LibraryGUIInfo.h [hunk 31]
```diff
@@ -47,6 +47,7 @@ class CLibraryGUIInfo : public CGUIInfoProvider
   mutable int m_libraryHasMovieSets;
   mutable int m_libraryHasSingles;
   mutable int m_libraryHasCompilations;
+  mutable int m_libraryHasBoxsets;
```
[the remark this answers is not in the document]
Comment (DaveTBlake) [reply xbmc/guilib/guiinfo/LibraryGUIInfo.h]: This is better left consistent with all the other library bools

**Review by DaveTBlake — APPROVED** [review]
Thanks for the skin related tweak, this looks really good @the-black-eagle especially as a first PR. 
I will kick off one more test build for @scott967 , but I am happy to see this merged.

**Comment by DaveTBlake** [comment]
jenkins build this please

**Pull request merged**